### PR TITLE
Introducing Block

### PR DESCRIPTION
**Pull request by marcromeyn**
### Goals :soccer:
First PR for the new PyTorch backend, this PR introduces the following user-facing classes:
- `Block`
- `Batch`
- `Sequence`

And internal classes:
- `TorchScriptWrapper`
- `BlockContainer`

### Implementation Details :construction:
`TorchScriptWrapper` makes it easier to work with torch-script. It will check the forward method and then stores various properties of it as constants. This way, torchscript doesn't need to worry about it.

### Testing Details :mag:
Since these are some of the lowest-level classes, I tried to maximize test-coverage as much as possible

CleanShot 2023-05-12 at 13 47 22@2x

### CODE REVIEW

File: merlin/models/torch/container.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        Parameters
+        ----------
+        module : nn.Module
+            The PyTorch module to be prepended.
+
+        Returns
+        -------
+        self
+        """
+        return self.insert(0, module)
```
Comment (edknv) [inline merlin/models/torch/container.py:75]: Should we wrap the `module` here with `self.wrap_module`?

File: merlin/models/torch/container.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        self.values[idx] = module
+
+        return self
+
+    def __delitem__(self, idx: Union[slice, int]) -> None:
+        self.values.__delitem__(idx)
+
+    def __add__(self, other) -> "BlockContainer":
+        for module in other:
+            self.append(module)
```
Comment (edknv) [inline merlin/models/torch/container.py:141]: Do we need to wrap the `module` here with `self.wrap_module`?

File: merlin/models/torch/utils/torchscript_utils.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            An optional batch object.
+
+        Returns
+        -------
+        torch.Tensor or Dict[str, torch.Tensor]
+            The output of the model.
+        """
+
+        if self.accepts_batch:
+            if self.requires_batch:
```
[the remark this answers is not in the document]
Comment (marcromeyn) [reply merlin/models/torch/utils/torchscript_utils.py:86]: It's so that if `requires_batch` is false, we cast `Optional[Batch]` into `Batch` and throw an RuntimeException if it's None.

File: merlin/models/torch/batch.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        if targets is None:
+            _targets = {}
+        elif torch.jit.isinstance(targets, Dict[str, torch.Tensor]):
+            _targets = targets
+        else:
+            raise ValueError("Targets must be a tensor or a dictionary of tensors")
+        self.targets: Dict[str, torch.Tensor] = _targets
+        self.sequences: Optional[Sequence] = sequences
+
+    def replace(
```
[the remark this answers is not in the document]
Comment (marcromeyn) [reply merlin/models/torch/batch.py:126]: Similar to the replace method of a `dataclass`. We still need to iron out the details but one potential use-case is to allow the `pre` in `Model` to transform the `Batch`. For instance to set the `sequences` (by doing padding + making). Another option is to do that transformation as part of the `Trainer` class.

File: merlin/models/torch/container.py (hunk 1 above)
Comment (marcromeyn) [reply merlin/models/torch/container.py:75]: Right now the wrapping of the module happens inside the `insert` method. This might be confusing. Do you think it's cleaner if we do it in 2 places? (The 2nd time that would amount to a no-op since it's already wrapped).

File: merlin/models/torch/container.py (hunk 2 above)
Comment (marcromeyn) [reply merlin/models/torch/container.py:141]: Right now it happens inside `self.append`.

File: merlin/models/torch/container.py (hunk 1 above)
Comment (edknv) [reply merlin/models/torch/container.py:75]: On second look, this looks fine. I think I confused `self.insert` with `self.values.insert`.

File: merlin/models/torch/container.py (hunk 2 above)
Comment (edknv) [reply merlin/models/torch/container.py:141]: Same with `self.insert`. This looks fine on second look.

**Review by edknv — APPROVED** [review]
